File: Kth_Smallest_Element_Sorted_Matrix.py
Time Complexity : O(nlogn)- where n comes from iterating the rows to find count and logn for binary search
Space Complexity : O(1) 
Did this code successfully run on Leetcode : Yes
Any problem you faced while coding this : No


# Binary Search Method
class Solution:
    def kthSmallest(self, matrix: List[List[int]], k: int) -> int:
        if not matrix:
            return 0
        
        n = len(matrix)
        low = matrix[0][0]
        high = matrix[n-1][n-1]
        print(n, low, high)
        while low < high:
            mid = (low + high) // 2
            print(mid)
            count = self.helper(matrix, mid, n)
            if count < k:
                low = mid + 1
            else:
                high = mid
                
        return low
    
    def helper(self, matrix, mid, n):
        count = 0
        j = n-1
        for i in range(n):
            while j >= 0 and matrix[i][j] > mid:
                j -= 1
            count = count + j + 1
        return count
        

File: KthSmallestElementSortedMatrix.cpp
/*
Intuition: The idea here is to apply binary search to get a mid value. We can then count the number elements 
upto that mid value and increase the low or high accordingly.

Time Complexity: O((N×log(Max−Min)))
Space Complexity: O(1)

*/
class Solution {
public:
    int kthSmallest(vector<vector<int>>& matrix, int k) {
        int n = matrix.size();
        int low = matrix[0][0];
        int high = matrix[n-1][n-1];
        while ( low < high){
        
            int mid = ( low) + ( high - low)/2;
            int count = getCount(matrix, mid);
            if ( count < k) low = mid+1;
            else high = mid;
        
        }
        return low;
    }
    int getCount(vector<vector<int>>& matrix,int mid){
        int count = 0;
        int j = matrix.size()-1;
        for ( int i = 0;  i < matrix.size(); i++){
            while ( j>=0 and matrix[i][j] > mid ) j--;
            count+=j +1;
        }
        
        
        return count;
    }
};